CacheManager.StackExchange.Redis.SelfCleaning.Tests/SelfCleaningRedisIntegrationTests.cs

```diff
@@ -0,0 +1,247 @@
+                .Callback<RedisKey, CommandFlags>((key, flags) => _fauxDatabase.Remove(key));
+        }
+
+        private void SetupConnectionMock()
+        {
+            _connectionMock.SetupGet(connection => connection.Configuration).Returns("connectionString");
+            _connectionMock.Setup(connection => connection.GetEndPoints(It.IsAny<bool>())).Returns(new EndPoint[1]);
+            _connectionMock.Setup(connection => connection.GetServer(It.IsAny<EndPoint>(), It.IsAny<object>()))
+                .Returns(_serverMock.Object);
+            _connectionMock.Setup(connection => connection.GetDatabase(It.IsAny<int>(), It.IsAny<object>()))
+                .Returns(_databaseMock.Object);
+        }
+
+        private void SetupCache()
+        {
+            double cleanupInterval = 100;
+            var cleanupTimer = new DefaultTimer(cleanupInterval);
+            _timeToLive = TimeSpan.FromSeconds(1);
+
+            _cache = CacheFactory.Build<DummyModel>(part => part
+                .WithProtoBufSerializer()
+                .WithSelfCleaningRedisConfiguration(_connectionMock.Object, cleanupTimer, _timeToLive,
+                    out string configurationKey)
+                .WithSelfCleaningRedisCacheHandle(configurationKey));
+
+            _onRemoveByHandleInvocations = new List<(CacheItemRemovedEventArgs, DateTime)>();
+            _cache.OnRemoveByHandle += (sender, args) => _onRemoveByHandleInvocations.Add((args, DateTime.Now));
+
+            _cache.CacheHandles.OfType<IStartable>().Single().Start();
+        }
+
+        #endregion
+
+        [TearDown]
+        public void TearDown()
+        {
+            _fauxDatabase.Clear();
+            _onRemoveByHandleInvocations.Clear();
+        }
+
+        [OneTimeTearDown]
+        public void OneTimeTearDown()
+        {
+            _cache.Dispose();
+        }
+
+        [TestCase(1, false)]
+        [TestCase(2, true)]
+        [TestCase(3, false)]
+        [TestCase(4, true)]
+        [TestCase(5, false)]
+        public void Put_AddSomeAmountOfKeys_OnRemoveByHandleCalledAfterTimeToLiveWithCorrectValueAndReason(
+            int amountOfKeys, bool delayBetweenInsertions)
+        {
+            // Arrange
+            IEnumerable<(string, DummyModel)> keys = Enumerable.Repeat("key", amountOfKeys)
+                .Select((key, i) => (key + i, new DummyModel {Property = "property" + i}));
+
+            TimeSpan delayBetweenInsertionsSpan = delayBetweenInsertions ? _timeToLive / 2 : TimeSpan.Zero;
+
+            // Act
+            var keysWithInsertionTimes = new List<(string, DummyModel, DateTime)>();
+
+            foreach ((string key, DummyModel value) in keys)
+            {
+                _cache[key] = value;
+                DateTime insertionTime = DateTime.Now;
+                keysWithInsertionTimes.Add((key, value, insertionTime));
+
+                Task.Delay(delayBetweenInsertionsSpan).Wait();
+            }
```

Comment:
Isnt that prettier?

![image](https://user-images.githubusercontent.com/62211836/78127294-88b9b680-741c-11ea-9f1b-f67e6de3beda.png)

Your call

Reply:
changed